### PR TITLE
pkg/native: added utils to make custom pkg easier

### PR DESCRIPTION
**Pull request by morlay**
When we want to create some custom tool base on cuelang. like a cuelang based [Tanka](http://tanka.dev/)

We need to add custom functions for this tool. 

important usage is the  custom formats, which cue not provided now,
for k8s configmap, like https://jsonnet.org/ref/stdlib.html#manifestIni did

instead of using `*internal.Package`, we could use pure go to registry custom packages.

```golang
package custom

import "cuelang.org/go/pkg/native"

func init() {
	native.Register(custom{})
}

type custom struct {
}

// pkg import path
func (custom) ImportPath() string {
	return "extension/custom"
}

// register as const values, when single value return
func (custom) StringConst() string {
	return "STRING"
}

// register as function, when need parameters, and returns value and error
func (custom) CustomFormat(v map[string]interface{}) ([]byte, error) {
	return json.Marshal(v)
}

// struct support too, and converting rules follow rules `encoding/json`
func (custom) CustomFormatWithStruct(v struct{ Name string `json:"name"` }) ([]byte, error) {
	return json.Marshal(v)
}
```

all inputs and outputs will be converted to cue.Value

Related: https://github.com/cuelang/cue/discussions/446

### CODE REVIEW

**Comment by verdverm** [comment]
For reference, this topic came up about a year ago. https://github.com/cuelang/cue/issues/350  (there was more in slack)

The TL;DR is that there is concern about fracturing the ecosystem.

If you add custom builtins, then you cannot read or process your Cue with the Cue CLI (reference implementation). This in effect creates dialects that only your tool can process. This may be ok for you and your tool, however, were this to happen with many tools, then people see examples of "Cue" from these many tools, ( each with their own dialect ), then people get confused about "Why does this Cue that I see over there does not work here? Why is the Cue CLI telling me this doesn't exist?" So if you make a tool that creates a dialect of Cue, it might not really be Cue anymore depending on who you ask.

There are a few existing capabilities and proposals that may make help you:

- The `text/template` package means you can create any file in any format or language
- `cue mod` will make dependencies and their imports really easy
- The "function" sugar issue (proposal?) that you already linked to
- Open an issue to support the `ini` format and see where that is said about that (not familiar enough with the complexities to comment myself)

**Comment by morlay** [comment]
@verdverm I know your concerns. 
sometimes, we need to create custom tool to make our work easier, before offical provided.

I'm not care how many tools i have to use, but if I want to bring a new good thing (cuelang is cool) to my team.
I need to make sure it could simple enough, and a single binary will be the best.

Recently, I'm bringing my team to get away from helm, 
starts with using `jsonnnet/tanka`, but without type constraints, jsonnet can't not stop people make mistakes.

So i start to create a new tool to solve my problems base the cuelang - https://github.com/octohelm/cuemod

Totally copy features from go modules. 
Instead of throughing `go get` to download. 
`cuemod` could download deps, and save the deps info into `mod.cue`. 

With adding new life cycle before mod load, 
the registered [extractors](https://github.com/octohelm/cuemod/tree/main/pkg/extractor) could automately extract from source codes to cue codes.

I hope the workflow could be easy like: 
```shell
mkdir -p ./demo && cd ./demo

cat << EOT > kube.cue
package kube

import (
   apps_v1 "k8s.io/api/apps/v1"
)

deployment: [string]: apps_v1.#Deployment

_labels: { "app": "nginx" }

deployment: nginx: spec: selector: matchLabels: _labels
deployment: nginx: spec: template: metadata: labels: _labels
deployment: nginx: spec: template: spec: {
	containers: [{
		name: "nginx"
		image: "nginx:1.11.10-alpine"
	}]
}
EOT
 
cuem eval ./kube.cue
```

And now, I have lots of helm charts and jsonnet. I can't refactor all of them to cue.

**JSONNET**: https://github.com/octohelm/cuemod/blob/main/pkg/cuemod/testdata/b/jsonnet_demo/jsonnet.cue
**Helm Charts**: https://github.com/octohelm/cuemod/blob/main/pkg/cuemod/testdata/b/helm_simple/local-path-provisioner.cue

This why the PR here. I really need add custom pkg to make it works in cue before i refactored.
`text/template` can't resolve it.

using `Attributes` to , but I don't know how to use Attributes to convert other format data to cue values.

**Comment by morlay** [comment]
@verdverm 

I think i could write a custom MarshalJSON to replace the cue.Value which with special tag.
to avoid to added custom functions

```cue
package kube

import "encoding/json"

config_map: xxx: spec: data: "xxx.toml":  json.Marshal({}) @manifest("toml")
config_map: xxx: spec: data: "xxx.ini":  json.Marshal({}) @manifest("ini")

app_from_helm: { 
    @helm() 
    
    values: {} 
    files: {} 
} 

app_from_jsonnet: { 
    @jsonnet()
    
    data: ''
    imports: {} 
}
```

But the decl attrs. seems have to wait https://github.com/cuelang/cue/pull/529

**Comment by myitcv** [comment]
@morlay thanks for starting a discussion on this. Generally speaking it's preferable to open a GitHub issue/discussion before a PR/Gerrit change (the Go project follows a similar model). PRs/Gerrit changes can easily follow later, once we have agreed and fleshed out a design and direction. If there isn't an issue that corresponds well to what you're after, please feel free to open a new one. Thanks
